slicerator.py

```diff
@@ -364,6 +360,80 @@ def _index_generator(new_indices, old_indices):
                 continue
 
 
+class Pipeline(object):
+    def __init__(self, ancestor, proc_func, propagate_attrs=None):
+        """A class to support lazy function evaluation on an iterable
+
+        Parameters
+        ----------
+        ancestor : object
+        proc_func : function
+            function that processes data returned by Slicerator. The function
+            acts element-wise and is only evaluated when data is actually
+            returned
+        """
+        # when list of propagated attributes are given explicitly,
+        # take this list and ignore the class definition
+        if propagate_attrs is not None:
+            self._propagate_attrs = propagate_attrs
+        else:
+            # check propagated_attrs field from the ancestor definition
+            self._propagate_attrs = []
+            if hasattr(ancestor, '_propagate_attrs'):
+                self._propagate_attrs += ancestor._propagate_attrs
+            if hasattr(ancestor, 'propagate_attrs'):
+                self._propagate_attrs += ancestor.propagate_attrs
+
+            # add methods having the _propagate flag
+            for attr in _iter_attr(ancestor):
+                if hasattr(attr, '_propagate_flag'):
+                    self._propagate_attrs.append(attr.__name__)
+
+        self._ancestor = ancestor
+        self._proc_func = proc_func
+
+    def _get(self, key):
+        # We need to copy here: else any _proc_func that acts inplace would
+        # change the ancestor value.
+        return self._proc_func(copy(self._ancestor[key]))
+
+    def __repr__(self):
+        msg = "{0} processed through {1}. Original repr:\n".format(
+                type(self._ancestor).__name__, self._proc_func.__name__)
+        old = '\n'.join("    " + ln for ln in repr(self._ancestor).split('\n'))
+        return msg + old
+
+    def __len__(self):
+        return self._ancestor.__len__()
+
+    def __iter__(self):
+        return (self._get(i) for i in range(len(self)))
+
+    def __getitem__(self, i):
+        """for data access"""
+        indices, new_length = key_to_indices(i, len(self))
+        if new_length is None:
+            return self._get(indices)
+        else:
+            return Slicerator(self, indices, new_length, self._propagate_attrs)
+
+    def __getattr__(self, name):
+        # to avoid infinite recursion, always check if public field is there
+        if '_propagate_attrs' not in self.__dict__:
+            self._propagate_attrs = []
+        if name in self._propagate_attrs:
+            return getattr(self._ancestor, name)
+        raise AttributeError
+
+    def __getstate__(self):
```

Comment:
Indeed, this deserves a better solution, but I also don't see a better way. `Pipeline` and `Slicerator` are (in the case of pims) essentially wrappers around file handles, which cannot be pickled.